spring-cloud-stream-binder-kafka-core/src/main/java/org/springframework/cloud/stream/binder/kafka/provisioning/KafkaTopicProvisioner.java

```diff
@@ -58,18 +69,26 @@
 public class KafkaTopicProvisioner implements ProvisioningProvider<ExtendedConsumerProperties<KafkaConsumerProperties>,
 		ExtendedProducerProperties<KafkaProducerProperties>>, InitializingBean {
 
+	private static final int DEFAULT_OPERATION_TIMEOUT = 30;
+
 	private final Log logger = LogFactory.getLog(getClass());
 
-	private final KafkaBinderConfigurationProperties configurationProperties;
+	private KafkaBinderConfigurationProperties configurationProperties;
 
-	private final AdminUtilsOperation adminUtilsOperation;
+	private final AdminClient adminClient;
 
 	private RetryOperations metadataRetryOperations;
 
+	private int operationTimeout = DEFAULT_OPERATION_TIMEOUT;
+
 	public KafkaTopicProvisioner(KafkaBinderConfigurationProperties kafkaBinderConfigurationProperties,
-								AdminUtilsOperation adminUtilsOperation) {
+								KafkaProperties kafkaProperties) {
+		Assert.isTrue(kafkaProperties != null, "KafkaProperties cannot be null");
+		Map<String, Object> stringObjectMap = kafkaProperties.buildAdminProperties();
+		//override if there is a binder level bootstrap server config or if it is different from base properties.
+		stringObjectMap.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaBinderConfigurationProperties.getKafkaConnectionString());
```

Comment:
This defaults to `localhost:9092` - that is bad because it will override the boot property.

We need to implement a hierarchy - only use the binder stuff if the properties are actually set.

You might need to look at https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/commit/6a31e9c94f604f7054dd4f50f905da4d8edd5c78 to see how @ilayaperumalg implemented it for the binder itself.

Reply:
Maybe add `getDefaultKafkaConnectionString()` to the properties.

Then

```java
if (ObjectUtils.isEmpty(props.get(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG))
         || !this.configurationProperties.getKafkaConnectionString().equals(this.configurationProperties.getDefaultKafkaConnectionString())) {
	props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, this.configurationProperties.getKafkaConnectionString());
}

The only case that won't work for is if there is a boot property and they explicitly want to connect to `localhost:9092` - but that's rather unlikely.